01/01c_elevator_repair.Rmd

```diff
@@ -0,0 +1,242 @@
+### Load the data
+
+```{r}
+failures <- read.csv('elevator_failures.csv')
+failures
+```
+
+### Build the model 
+
+Let's assume that the elevator failures happen at a constant probability per time of happening and are uncorrelated with each other- so they'll be described reasonbly well by a homogeneous Poisson process.
+
+The Poisson likelihood has one parameter ($\lambda$) that is equal to both the mean and the standard deviation of the distribution. $\lambda$ has to be nonnegative but doesn't have an upper bound, so we'll use a Gamma distribution.
+
+Prior:
+
+$\lambda \sim gamma(a,b)$
+
+Likelihood:
+
+$y \sim Poisson(\lambda)$
+
+**Note:** <span style="color:red">rgamma uses (shape, scale) to parameterize Gamma distribution; stan uses (alpha, beta) where alpha=shape and beta=1/scale.</span> Example samples from a Gamma distribution:
+
+```{r, plot-prior}
+a <- 2
+b <- 2
+prior_samples <- rgamma(n=1000,shape=a,scale=b)
+hist(prior_samples,100)
+```
+
+Now build the stan model:
+
+```{r}
+model_code = "
+data {
+    int<lower=0> N;
+    int<lower=0> y[N];
+}
+parameters {
+    real<lower=0> lam;
+}
+model {
+    lam ~ gamma(2,0.5); // SPECIFY YOUR PRIOR HYPERPARAMETERS HERE
+    y ~ poisson(lam);
+}
+"
+```
+
+```{r, results="hide"}
+model <- stan_model(model_code=model_code)
+```
+
+### Draw samples from the posterior
+
+Run inference on the data from your building.
+
+```{r}
+data <- list(
+    N=nrow(failures),
+    y=failures$you
+)
+```
+
+```{r, results='hide'}
+fit <- sampling(object=model, data=data, iter=10000,chains=4)
+fit
+```
+
+```{r}
+posterior_samples <- extract(fit)$lam
+dim(posterior_samples)
+```
+
+```{r, plot-posterior}
+# creating a dataframe to nicely use ggplot
+prior_post <- data.frame('sample'=c(rep('prior',length(prior_samples)),rep('posterior',length(posterior_samples))),
+                         'value'=c(prior_samples,posterior_samples))
+g <- ggplot(prior_post, aes(x=value,stat(density),fill=sample))
+g + geom_histogram(bins=100, position='identity',alpha=0.7) + 
+    theme(legend.justification=c(1,1), legend.position=c(1,1), legend.title=element_blank())
+```
+
+### Find the mean and 90% credible interval of the average number of failures per year¶
+
+```{r}
+mean(posterior_samples)
+```
+
+```{r}
+paste("95 percent:", quantile(posterior_samples,0.95))
+paste("5 percent:", quantile(posterior_samples, 0.05))
+```
+
+
+### Challenge: use your model to make an actual decision
+
+Let's say that the elevator company sells service contracts for repairing broken elevators:
+
+* You can prepay for $N$ service visits per year, at a cost of $N \times \$1000$ (whether or not you use them all)
+* Non-prepaid service visits cost \$1500
+
+**Use your model to predict the value of $N$ that will minimize expected cost**
+
+Suggested approach: write a function that inputs the number of prepaid visits and the number of failures for a given year, and outputs the total cost.
+
+```{r}
+cost_function <- function(num_failures,n){
```

Comment:
Starting here I have some suggestions on how to re-write the rest of these code blocks to be a little more R friendly. There are certain techniques that work really well in R (ex. vectorization) and don't rely as much on for loops and appending values to lists. But I think I'll save them for a separate pull-request because I don't want to put too much here. Plus they don't change any of the answers so I'll let you decide if they are worth to merge.